### PR TITLE
Support using mmap to load data 

### PR DESCRIPTION
**Pull request by yah01**
Signed-off-by: yah01 <yang.cen@zilliz.com>
/kind improvement

This also avoids coping when matching prefix with string index,
also avoids coping when retrieving string column,
also stores strings continuously in memory to improve locality,
also fixes the data race when updating row count of sealed segment.

This may break the compatibility with Windows

### CODE REVIEW

File: internal/querynode/segment_loader.go [hunk 1]
```diff
@@ -913,7 +913,9 @@ func (loader *segmentLoader) checkSegmentSize(collectionID UniqueID, segmentLoad
 				usedMemAfterLoad += neededMemSize
 				usedLocalSizeAfterLoad += neededDiskSize
 			} else {
-				usedMemAfterLoad += uint64(funcutil.GetFieldSizeFromFieldBinlog(fieldBinlog))
+				if fieldID != common.RowIDField && fieldID != common.TimeStampField {
+					usedMemAfterLoad += uint64(funcutil.GetFieldSizeFromFieldBinlog(fieldBinlog))
```
Comment (jaime0815) [inline internal/querynode/segment_loader.go:917]: Why exclude RowID filed and timestamp filed when calculating memory usage?

Comment (yah01) [reply internal/querynode/segment_loader.go:917]: This demo still keeps row ID and timestamp in memory entirely

**Comment by yah01** [comment]
It's hard to support varchar filter with new string storage format, will skip it first

**Comment by yah01** [comment]
fix #22106

File: internal/core/src/common/Utils.h [hunk 2]
```diff
@@ -166,4 +173,198 @@ MatchKnowhereError(knowhere::Status status) {
...
+    static int mmap_flags = MAP_SHARED;
+#ifdef MAP_POPULATE
+    // macOS doesn't support MAP_POPULATE
+    mmap_flags |= MAP_POPULATE;
+#endif
+    // Allocate memory
+    if (info.mmap_dir_path == nullptr) {
+        auto data_type = field_meta.get_data_type();
+        auto data_size = field_meta.get_sizeof() * info.row_count;
+        size_t size = datatype_is_string(data_type) ? GetStringsSize(info.field_data) : data_size;
```
Comment (xiaofan-luan) [inline internal/core/src/common/Utils.h:315]: I've seen many datatype_is_string check
Should we change to is_variable_type for futher extension

File: internal/core/src/index/StringIndex.h [hunk 3]
```diff
@@ -16,12 +16,14 @@
...
+
 #include <memory>
+#include <string>
 #include <vector>
+
 #include "index/Meta.h"
-#include <pb/schema.pb.h>
+#include "index/ScalarIndex.h"
 
 namespace milvus::index {
```
[the remark this answers is not in the document]
Comment (yah01) [reply internal/core/src/index/StringIndex.h:28]: For now the index owns the data, changing to string_view requires to additional data management, will try this later with optimizing string

**Comment by xiaofan-luan** [comment]
@yanliang567 mmap features in 2.3 is there

**Pull request merged**